src/Workspaces/CoreTest/UtilityTest/BKTreeTests.cs

```diff
@@ -0,0 +1,151 @@
+        {
+            string[] testValues = { "cook", "book", "books", "cake", "what", "water", "Cape", "Boon", "Cook", "Cart" };
+            var tree = BKTree.Create(testValues);
+
+            var results1 = tree.Find("wat", threshold: 1);
+            Assert.Single(results1, "what");
+
+            var results2 = tree.Find("wat", threshold: 2);
+            Assert.True(results2.SetEquals(Expected("cart", "what", "water")));
+
+            var results3 = tree.Find("caqe", threshold: 1);
+            Assert.True(results3.SetEquals(Expected("cake", "cape")));
+        }
+
+        [Fact]
+        public void PermutationTests()
+        {
+            string[] testValues = { "cook", "book", "books", "cake", "what", "water", "Cape", "Boon", "Cook", "Cart" };
+            TestTreeInvariants(testValues);
+        }
+
+        private void TestTreeInvariants(string[] testValues)
+        {
+            var tree = BKTree.Create(testValues);
+
+            foreach (var value in testValues)
+            {
+                // With a threshold of 0, we should only find exactly the item we're searching for.
+                var items = tree.Find(value, threshold: 0);
+                Assert.Single(tree.Find(value, threshold: 0), value.ToLower());
+            }
+
+            foreach (var value in testValues)
+            {
+                // With a threshold of 1, we should always at least find the item we're looking for.
+                // But we may also find additional items along with it.
+                var items = tree.Find(value, threshold: 1);
+                Assert.Contains(value.ToLower(), items);
+
+                // We better not be finding all items.
+                Assert.NotEqual(testValues.Length, items.Count);
+            }
+
+            foreach (var value in testValues)
+            {
+                // If we delete each individual character in each search string, we should still
+                // find the value in the tree.
+                for (var i = 0; i < value.Length; i++)
+                {
+                    var items = tree.Find(Delete(value, i), threshold: null);
+                    Assert.Contains(value.ToLower(), items);
+
+                    // We better not be finding all items.
+                    Assert.NotEqual(testValues.Length, items.Count);
+                }
+            }
+
+            foreach (var value in testValues)
+            {
+                // If we add a random character at any location in a string, we should still 
+                // be able to find it.
+                for (var i = 0; i <= value.Length; i++)
+                {
+                    var items = tree.Find(Insert(value, i, 'Z'), threshold: null);
+                    Assert.Contains(value.ToLower(), items);
+
+                    // We better not be finding all items.
+                    Assert.NotEqual(testValues.Length, items.Count);
+                }
+            }
+
+            foreach (var value in testValues)
```

Comment:
You've painstakingly enumerated most of the cases; can you add the "change a character" case?